crates/subspace-farmer/src/single_disk_farm/piece_cache.rs

```diff
@@ -101,28 +94,33 @@ impl DiskPieceCache {
     pub(crate) fn contents(
         &self,
     ) -> impl ExactSizeIterator<Item = (Offset, Option<PieceIndex>)> + '_ {
-        self.inner
-            .read_mmap
-            .array_chunks::<{ Self::element_size() }>()
-            .enumerate()
-            .map(|(offset, chunk)| {
-                let (piece_index_bytes, piece_bytes) = chunk.split_at(PieceIndex::SIZE);
-                let piece_index = PieceIndex::from_bytes(
-                    piece_index_bytes
-                        .try_into()
-                        .expect("Statically known to have correct size; qed"),
-                );
-                // Piece index zero might mean we have piece index zero or just an empty space
-                let piece_index = if piece_index != PieceIndex::ZERO
-                    || piece_bytes.iter().any(|&byte| byte != 0)
-                {
+        let file = &self.inner.file;
+        let mut element = vec![0; Self::element_size()];
+
+        (0..self.inner.num_elements).map(move |offset| {
+            if let Err(error) =
+                file.read_exact_at(&mut element, (offset * Self::element_size()) as u64)
+            {
+                warn!(%error, %offset, "Failed to read cache element #1");
+                return (Offset(offset), None);
+            }
+
+            let (piece_index_bytes, piece_bytes) = element.split_at(PieceIndex::SIZE);
+            let piece_index = PieceIndex::from_bytes(
+                piece_index_bytes
+                    .try_into()
+                    .expect("Statically known to have correct size; qed"),
+            );
+            // Piece index zero might mean we have piece index zero or just an empty space
+            let piece_index =
+                if piece_index != PieceIndex::ZERO || piece_bytes.iter().any(|&byte| byte != 0) {
```

Comment:
This is a good candidate for `then` or `then_some`

Reply:
Will address in the next PR :+1:

Reply:
I tried and it is less readable that way unfortunately

Reply:
probably, because you tried just to call `then` from the large condition. Try creating a separate boolean with a good name and then it looks more idiomatic rust.